optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/RandomListChangeIterator.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list;
+
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Random;
+import java.util.TreeMap;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.heuristic.move.Move;
+import org.optaplanner.core.impl.heuristic.selector.common.iterator.UpcomingSelectionIterator;
+
+/**
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ */
+public class RandomListChangeIterator<Solution_> extends UpcomingSelectionIterator<Move<Solution_>> {
+
+    private final ListVariableDescriptor<Solution_> listVariableDescriptor;
+    private final Random workingRandom;
+
+    private final NavigableMap<Integer, Object> sourceEntityByIndexNavigableMap;
+    private final int sourceIndexRange;
+    private final NavigableMap<Integer, Object> destinationEntityByIndexNavigableMap;
+    private final int destinationIndexRange;
+
+    public RandomListChangeIterator(
+            ListVariableDescriptor<Solution_> listVariableDescriptor,
+            List<Object> workingEntityList,
```

Comment:
If the entity list is filtered implicitly (for example pinning causes entitySelector filtering) or explicitly (in solver config, they added filtering for the entity selector) or the entitySelector is otherwise manipulated (SHUFFLED, mimics, ...),
it ignores that.

In the long term however, we have seen that not all combinations make sense to keep supporting (for example step ached SHUFFLED entitySelector under JIT random change move selection)

Reply:
Fixed. Replaced with an entitySelector.